src/tails/node.ts

```diff
@@ -0,0 +1,229 @@
+      `npm install --global ${packages.join(" ")}`,
+      engine.vars(this.frameworks)
+    );
+  }
+
+  installCommand(): string | null {
+    let npmCommand = "npm install";
+    for (const framework of this.frameworks) {
+      if (framework.installCommand) {
+        npmCommand = framework.installCommand;
+        break;
+      }
+    }
+    if (this.pkgMgr === "yarn") {
+      return npmCommand.replace(/npm i(nstall)?/, "yarn install");
+    }
+    return engine.interpolate(npmCommand, engine.vars(this.frameworks));
+  }
+
+  buildCommand(): string | null {
+    const scripts: string[] = [];
+    // Note: we assume build trumps all. If there's a build script we don't
+    // even run tsc. Though we should use the correct package manager to build.
+    if (this.scripts.includes("build")) {
+      scripts.push("npm run build");
+    } else {
+      if (this.lang === "ts-local") {
+        scripts.push("./node_modules/.bin/tsc");
+      } else if (this.lang === "ts-global") {
+        scripts.push("tsc");
+      }
+      for (const framework of this.frameworks) {
+        if (framework.buildCommand) {
+          scripts.push(framework.buildCommand);
+          break;
+        }
+      }
+    }
+
+    if (!scripts.length) {
+      return null;
+    }
+
+    let command = scripts.join(" && ");
+    if (this.pkgMgr === "yarn") {
+      command = command.replace("npm run", "yarn run");
+    }
+
+    return engine.interpolate(command, engine.vars(this.frameworks));
+  }
+
+  devCommand(): string | null {
+    let devCommand: string | null = null;
+    if (this.scripts.includes("dev")) {
+      devCommand = "npm run dev";
+    }
+    if (!devCommand && this.scripts.includes("start")) {
+      devCommand = "npm run start";
+    }
+    if (!devCommand) {
+      for (const framework of this.frameworks) {
+        if (framework.devCommand) {
+          devCommand = framework.devCommand;
+          break;
+        }
+      }
+    }
+    if (devCommand && this.pkgMgr === "yarn") {
+      return devCommand.replace("npm", "yarn");
+    }
+
+    return engine.interpolate(devCommand, engine.vars(this.frameworks));
+  }
+}
+
+/**
+ * Detects a node codebase and returns the Strategy for the codebase with the
+ * detected framework(s).
+ */
+export async function detect(fs: engine.FileSystem): Promise<engine.Strategy | null> {
+  let pkgJsonRaw: string | null = null;
+  let hasYarn = false;
+  let hasTsconfig = false;
+  await Promise.all([
+    (async () => {
+      try {
+        pkgJsonRaw = await fs.read("package.json");
+      } catch (err: any) {
+        if (err.code === "ENOENT") {
+          pkgJsonRaw = null;
+        }
+      }
+    })(),
+    (async () => {
+      hasYarn = await fs.exists("yarn.lock");
+    })(),
+    (async () => {
+      hasTsconfig = await fs.exists("tsconfig.json");
+    })(),
+  ]);
+  if (!pkgJsonRaw) {
+    return null;
+  }
+  // TODO: Find out why pkgJsonRaw is never. TypeScript isn't seeing the assignment
```

Comment:
You should destructure your Promise.all, `const [pkgJsonRaw, hasYarn, hasTsconfig] = Promise.all([ fs.read("package.json"), fs.exists("yarn.lock"), fs.exists("tsconfig.json")]);`

Reply:
That's how I tried to do it, but I think Promise.all got angry with me when it had nonhomogenous return types. I think whoever did the type annotation for Promise.all did `<Promise<T>[]>` and not `<T extends any[]>`, which would have captured tuples correctly.

Reply:
Found it. I was on the right track about template inference. If I add "as const" to the end of the array it turns it into a tuple and I can destructure. Hopefully this shit gets figured out in the new version of TypeScript when you can infer const.